R/method.R

```diff
@@ -1,145 +1,163 @@
+
+    if (length(signature) != 1 || class_type(signature[[1]]) != "r7") {
+      msg <- sprintf(
+        "When registering methods for S3 generic %s(), signature be a single R7 class",
+        generic_name
+      )
+      stop(msg, call. = FALSE)
+    }
+    class <- signature[[1]]@name
+    registerS3method(generic_name, class, method, envir = parent.frame())
+  } else {
+    check_method(method, signature, generic)
+    register_r7_method(generic, signature, method)
   }
 
-  unlist(lapply(names(x), function(class) get_all_methods(x[[class]], c(signature, class))), recursive = FALSE)
+  invisible()
 }
 
-as_signature <- function(signature) {
-  if (!is.list(signature)) {
-    signature <- list(signature)
-  }
+register_r7_method <- function(generic, signature, method) {
+  # Flatten out unions to individual signatures
+  signatures <- flatten_signature(signature)
 
-  for (i in seq_along(signature)) {
-    signature[[i]] <- as_class(signature[[i]], arg = sprintf("signature[[%i]]", i))
+  # Register each method
+  for (signature in signatures) {
+    method <- R7_method(method, generic = generic, signature = signature)
+    generic_add_method(generic, signature, method)
   }
-  signature
+
+  invisible()
 }
 
-method_impl <- function(generic, signature, ignore) {
-  out <- .Call(method_, generic, signature, ignore)
-  if (!is.null(out)) {
-    return(out)
-  }
+flatten_signature <- function(signature) {
+  # Unpack unions
+  sig_is_union <- vlapply(signature, is_union)
+  signature[sig_is_union] <- lapply(signature[sig_is_union], prop, "classes")
+  signature[!sig_is_union] <- lapply(signature[!sig_is_union], list)
 
-  # If no R7 method is found, see if there are any S3 methods registered
-  if (inherits(generic, "R7_generic")) {
-    args <- generic@dispatch_args
-    generic <- generic@name
-  } else {
-    args <- setdiff(names(formals(generic)), "...")
-    generic <- find_function_name(generic, topenv(environment(generic)))
-  }
+  # Create grid of indices
+  indx <- lapply(signature, seq_along)
+  comb <- as.matrix(rev(do.call("expand.grid", rev(indx))))
+  colnames(comb) <- NULL
 
-  if (length(signature) > 1) {
-    out <- getS3method(generic, s3_class_name(signature[[1]]), optional = TRUE)
-    if (!is.null(out)) {
-      return(out)
-    }
+  rows <- lapply(1:nrow(comb), function(i) comb[i, ])
+  lapply(rows, function(row) Map("[[", signature, row))
+}
+
+as_generic <- function(x) {
+  if (inherits(x, "R7_generic") || is_external_generic(x)) {
+    return(x)
   }
 
-  # If no method found check if the generic has a default method
-  out <- getS3method(generic, "default", optional = TRUE)
-  if (!is.null(out)) {
-    return(out)
+  if (!is.function(x)) {
+    msg <- sprintf("`generic` must be a function, not a %s", obj_desc(x))
+    stop(msg, call. = FALSE)
   }
 
-  method_lookup_error(generic, args, signature)
+  # For now, assume that it's an S3 generic
+  attr(x, "name") <- find_generic_name(x)
+  class(x) <- "R7_S3_generic"
```

Comment:
I feel like `structure(list(generic = x, name = name), class = "R7_S3_generic")` may be a little more structured? It feels a little wrong to me to add attributes and classes directly to the S3 generic